src/ui.js

```diff
@@ -77,29 +77,62 @@ function handleFileSelect(map, evt) {
 
     modal.show();
 
-    let openFile = file => new Promise(resolve => {
+    let handleGpx = file => new Promise(resolve => {
         let reader = new FileReader();
-        reader.onload = () => resolve(reader.result);
+        reader.onload = () => {
+            parseGPX(reader.result)
+                .then(parsedTracks => {
+                    parsedTracks.forEach(track => {
+                        track.filename = file.name;
+                        tracks.push(track);
+                        map.addTrack(track);
+                        modal.addSuccess();
+                    });
+                })
+                .then(resolve);
```

Comment:
Again, wrapped promise not needed: .then() returns a promise that resolves when it completes/when the promise its code returns resolves